tests/scripts/thread-cert/test_br_upgrade_router_role.py

```diff
@@ -0,0 +1,156 @@
+#
+# This test verifies behavior of Border Routers (which provide IP connectivity) requesting router role within Thread
+# mesh.
+
+# Topology:
+#
+#  5 FTD nodes, all connected.
+#  Topology is expected to change during tests.
+#
+
+LEADER = 1
+ROUTER = 2
+BR1 = 3
+BR2 = 4
+BR3 = 5
+
+
+class SrpAutoStartMode(thread_cert.TestCase):
+    USE_MESSAGE_FACTORY = False
+    SUPPORT_NCP = False
+
+    TOPOLOGY = {
+        LEADER: {
+            'name': 'leader',
+            'mode': 'rdn',
+        },
+        ROUTER: {
+            'name': 'reader',
+            'mode': 'rdn',
+        },
+        BR1: {
+            'name': 'br-1',
+            'mode': 'rdn',
+        },
+        BR2: {
+            'name': 'br-2',
+            'mode': 'rdn',
+        },
+        BR3: {
+            'name': 'br-3',
+            'mode': 'rdn',
+        },
+    }
+
+    def test(self):
+        leader = self.nodes[LEADER]
+        router = self.nodes[ROUTER]
+        br1 = self.nodes[BR1]
+        br2 = self.nodes[BR2]
+        br3 = self.nodes[BR3]
+
+        #-------------------------------------------------------------------------------------
+        # Set the router upgrade threshold to 2 on all nodes.
+
+        for node in [leader, router, br1, br2, br3]:
+            node.set_router_upgrade_threshold(2)
+
+        #-------------------------------------------------------------------------------------
+        # Start the leader and router.
+
+        leader.start()
+        self.simulator.go(10)
+        self.assertEqual(leader.get_state(), 'leader')
+
+        router.start()
+        self.simulator.go(125)
+        self.assertEqual(router.get_state(), 'router')
+
+        #-------------------------------------------------------------------------------------
+        # Start all three BRs, we expect all to stay as child since there are already two
+        # routers in the Thread mesh.
+
+        br1.start()
+        br2.start()
+        br3.start()
+        self.simulator.go(125)
+        self.assertEqual(br1.get_state(), 'child')
+        self.assertEqual(br2.get_state(), 'child')
+        self.assertEqual(br3.get_state(), 'child')
+
+        #-------------------------------------------------------------------------------------
+        # Add an external route on `br1`, it should now try to become a router requesting
+        # with status BorderRouterRequest. Verify that leader allows it to become router.
+
+        br1.add_route('2001:dead:beef:cafe::/64', stable=True)
+        br1.register_netdata()
+        self.simulator.go(15)
+        self.assertEqual(br1.get_state(), 'router')
+
+        #-------------------------------------------------------------------------------------
+        # Add a prefix with default route on `br2`, it should also become a router.
+
+        br2.add_prefix('2001:dead:beef:2222::/64', 'paros')
+        br2.register_netdata()
+        self.simulator.go(15)
+        self.assertEqual(br2.get_state(), 'router')
```

Comment:
how do we make sure the BR does not downgrade if the number of BRs exceed the `RouterDowngradeThreshold` ?

Reply:
If it downgrades (for some reason), we notice it from the `HandleNotifierEvent()` role change (role become `Child`) and can then send the request again if still eligible.

Reply:
@abtink Maybe I misunderstand your above comment. But if it downgrades (due to the usual downgrade rules for Routers), and then upgrades again (due to the new BorderRouterRequest rule), then the first thing it would do after upgrading is downgrading again, right? So an infinite sequence of upgrade/downgrade actions would result which isn't helpful.

I would say that if a BR is eligible to send BorderRouterRequest, but it is already a Router, then it should avoid downgrading.   So that would need to be checked in the regular downgrade code.

Reply:
I see. No I missed that.  I agree we should ensure to add a check in downgrade code path to not allow a BR that is eligible for such as request to be downgraded (this would still be limited to two such BRs in network data).

Reply:
Added such a check in the new push (it is for now a separate commit to make it easier to read/review).